src/main/scala/akka/stream/contrib/ZipInputStreamSource.scala

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import java.util.zip.{ ZipEntry, ZipInputStream }
+import akka.stream.Attributes.{ InputBuffer, name }
+import akka.stream.contrib.ZipInputStreamSource.{ DefaultAllowedZipExtensions, DefaultChunkSize, ZipEntryData }
+import akka.stream.impl.Stages.DefaultAttributes.IODispatcher
+import akka.stream.scaladsl.Source
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, OutHandler }
+import akka.stream.{ Attributes, Outlet, SourceShape }
+import akka.util.ByteString
+import akka.util.ByteString.ByteString1C
+import scala.annotation.tailrec
+import scala.collection.immutable
+import scala.concurrent.{ Future, Promise }
+import scala.util.control.NonFatal
+
+/**
+ * This companion defines a factory for [[ZipInputStreamSource]] instances,
+ * see [[ZipInputStreamSource.apply]].
+ */
+object ZipInputStreamSource {
+
+  final val DefaultChunkSize = 8192
+  final val DefaultAllowedZipExtensions = immutable.Seq(".zip")
+
+  /**
+   * Data type for zip entries.
+   *
+   * @param name file name
+   * @param creationTime The last modification time of the entry in milliseconds
+   *          since the epoch, or -1 if not specified
+   */
+  case class ZipEntryData(name: String, creationTime: Long)
+
+  /**
+   * Factory for [[ZipInputStreamSource]] instances wrapped
+   * into [[Source]].
+   *
+   * @param in a function that builds a [[ZipInputStream]]
+   * @param chunkSize the size of the chunks
+   * @param allowedZipExtensions collection of allowed extensions for zipped containers,
+   *                    as zip or jar files. Only ".zip" extension allowed by default.
+   * @return [[ZipInputStreamSource]] instance
+   */
+  def apply(
+    in:                   () => ZipInputStream,
+    chunkSize:            Int                   = DefaultChunkSize,
+    allowedZipExtensions: immutable.Seq[String] = DefaultAllowedZipExtensions
+  ): Source[(ZipEntryData, ByteString), Future[Long]] =
```

Comment:
Yeah, exactly. That way the entries would be streams themself instead of unbounded `ByteString`s.

Take a look at [StreamOfStreams.scala](https://github.com/akka/akka/blob/03395d5739ee371e518331f66dfd84447a8ff9ce/akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala) for inspiration.